intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/staticanalysis/IntelliJMarkupModelListener.java

```diff
@@ -0,0 +1,220 @@
+        this.generatedWarnings.clear();
+
+        addRemovedWarnings(this.trackingEventManager, this.warnings.stream().map(IntelliJMarkupModelListener::classifyWarning), this.document);
+        this.warnings.clear();
+    }
+
+    @Override
+    public void afterAdded(@NotNull RangeHighlighterEx rangeHighlighterEx) {
+        if (isWarningRangeHighlighter(rangeHighlighterEx)) {
+            final DateTime creationTime = DateTime.now();
+
+            this.timeMapping.put(rangeHighlighterEx, creationTime);
+            this.generatedWarnings.add(new Warning<>(rangeHighlighterEx, getLineNumberForHighlighter(rangeHighlighterEx), creationTime.toDate()));
+        }
+    }
+
+    @Override
+    public void beforeRemoved(@NotNull RangeHighlighterEx rangeHighlighterEx) {
+        DateTime creationTime = this.timeMapping.remove(rangeHighlighterEx);
+
+        if (isWarningRangeHighlighter(rangeHighlighterEx)) {
+            // Only process a deletion if we hadn't encountered this marker in this session before.
+            // If we did encounter it, remove returns `true` and the warning is not saved as removed.
+            if (!this.generatedWarnings.removeIf(warning -> warning.type.equals(rangeHighlighterEx))) {
+                this.warnings.add(createWarningFromRangeHighlighter(rangeHighlighterEx, creationTime));
+            }
+        }
+    }
+
+    @Override
+    public void attributesChanged(@NotNull RangeHighlighterEx rangeHighlighterEx, boolean b, boolean b1) {
+    }
+
+    @Override
+    public void dispose() {
+    }
+
+    private static boolean isWarningRangeHighlighter(@NotNull RangeHighlighterEx rangeHighlighterEx) {
+        return rangeHighlighterEx.getLayer() == HighlighterLayer.WARNING;
+    }
+
+    @NotNull
+    private Warning<RangeHighlighterEx> createWarningFromRangeHighlighter(@NotNull RangeHighlighterEx rangeHighlighterEx, DateTime creationTime) {
+        final DateTime now = DateTime.now();
+
+        return new Warning<>(
+                rangeHighlighterEx,
+                getLineNumberForHighlighter(rangeHighlighterEx),
+                now.toDate(),
+                creationTime == null ? -1 : Seconds.secondsBetween(creationTime, now).getSeconds()
+        );
+    }
+
+    private int getLineNumberForHighlighter(RangeHighlighterEx warning) {
+        return intellijDocument.getLineNumber(warning.getAffectedAreaStartOffset());
+    }
+
+    private static Warning<String> classifyWarning(Warning<RangeHighlighterEx> warning) {
+        return new Warning<>(classifyWarningTypeFromHighlighter(warning.type), warning.lineNumber, warning.warningCreationTime, warning.secondsBetween);
+    }
+
+    @NotNull
+    private static String classifyWarningTypeFromHighlighter(@NotNull RangeHighlighterEx rangeHighlighterEx) {
+        final Object errorStripeTooltip = rangeHighlighterEx.getErrorStripeTooltip();
```

Comment:
Does this really need to be final?

Reply:
They are never changed nor should they, so this is mostly a hint to the compiler for optimization purposes.

Reply:
I would prefer "final" to be removed if it is not absolutely necessary to compile. It burdens people's mind as to why it can't just be a normal local variable.